distributed/shuffle/tests/test_shuffle.py

```diff
@@ -469,3 +475,279 @@ async def test_clean_after_close(c, s, a, b):
+@pytest.mark.parametrize("n_input_partitions", [1, 2, 10])
+@pytest.mark.parametrize("npartitions", [1, 20])
+@pytest.mark.parametrize("barrier_first_worker", [True, False])
+@gen_test()
+async def test_basic_lowlevel_shuffle(
+    tmpdir,
+    loop_in_thread,
+    n_workers,
+    n_input_partitions,
+    npartitions,
+    barrier_first_worker,
+):
+    dfs = []
+    rows_per_df = 10
+    for ix in range(n_input_partitions):
+        df = pd.DataFrame({"x": range(rows_per_df * ix, rows_per_df * (ix + 1))})
+        df["_partition"] = df.x % npartitions
+        dfs.append(df)
+
+    workers = list("abcdefghijklmn")[:n_workers]
+
+    worker_for_mapping = {}
+
+    for part in range(npartitions):
+        worker_for_mapping[part] = get_worker_for(part, workers, npartitions)
+    assert len(set(worker_for_mapping.values())) == min(n_workers, npartitions)
+    schema = pa.Schema.from_pandas(dfs[0])
+
+    local_shuffle_pool = ShuffleTestPool()
+    shuffles = []
+    for ix in range(n_workers):
+        shuffles.append(
+            local_shuffle_pool.new_shuffle(
+                name=workers[ix],
+                worker_for_mapping=worker_for_mapping,
+                schema=schema,
+                directory=tmpdir,
+                loop=loop_in_thread,
+            )
+        )
+    random.seed(42)
+    if barrier_first_worker:
+        barrier_worker = shuffles[0]
+    else:
+        barrier_worker = random.sample(shuffles, k=1)[0]
+
+    try:
+        for ix, df in enumerate(dfs):
+            s = shuffles[ix % len(shuffles)]
+            s.add_partition(df)
+            await asyncio.sleep(0.05)
+
+        await barrier_worker.barrier()
+
+        def _done():
+            return [s.done() for s in shuffles]
+
+        assert sum(_done()) == max(0, n_workers - npartitions)
+
+        all_parts = []
+        for part, worker in worker_for_mapping.items():
+            s = local_shuffle_pool.shuffles[worker]
+            all_parts.append(s.get_output_partition(part))
+
+        all_parts = await asyncio.gather(*all_parts)
+
+        df_after = pd.concat(all_parts)
+        assert all(_done())
+    finally:
+        await asyncio.gather(*[s.close() for s in shuffles])
+    assert len(df_after) == len(pd.concat(dfs))
+
+
+@gen_test()
+async def test_slow_offload(tmpdir, loop_in_thread):
+    # Ensure that all data is
+    dfs = []
+    rows_per_df = 10
+    n_input_partitions = 2
+    npartitions = 2
+    for ix in range(n_input_partitions):
+        df = pd.DataFrame({"x": range(rows_per_df * ix, rows_per_df * (ix + 1))})
+        df["_partition"] = df.x % npartitions
+        dfs.append(df)
+
+    workers = ["A", "B"]
+
+    worker_for_mapping = {}
+    partitions_for_worker = defaultdict(list)
+
+    for part in range(npartitions):
+        worker_for_mapping[part] = w = get_worker_for(part, workers, npartitions)
+        partitions_for_worker[w].append(part)
+    schema = pa.Schema.from_pandas(dfs[0])
```

Comment:
This is all boilerplate I hope we can get rid of eventually. Right now not a prio